### DISCUSSION TITLE
standalone PMD performance optimization

### DISCUSSION THREAD

**Discussion by caesarkim** [pmd/pmd#3019] [Q&A]
I am comparing the standalone PMD (downloaded from https://pmd.github.io/) to the PMD gradle plugin, because there seems to be a big difference in terms of performance.  I have a little over 200 osgi bundles and a small number of rules that I am applying. I have about 5 rules in total. 

When I run pmd gradle plugin(https://docs.gradle.org/current/userguide/pmd_plugin.html) in a gradle environment (on Windows 10, i7 2.6GH) with parallel execution enabled against 220 bundles, it takes about 10 minutes without compileJava and  test tasks. I am excluding compileJava, testOSGi, test tasks which are done in a separate step. 

However, when I use standalone PMD against the same number of bundles with the same number of 5 rules, it takes only 2 minutes. I set all the jars generated from the build to the auxclasspath so that PMD resolves types. 

here is the command for the gradle 
`./gradlew --build-file build.gradle  pmdMain  -x classes -x testClasses -x build -x check -x compileJava -x compileTestJava -x jar -x test -x testOSGI`

PMD command 
`pmd.bat  -dir C:/Projects/bundles -encoding UTF-8 -failOnViolation false -R C:/Projects/pmd/rules/CustomJavaRules.xml  -cache C:/Projects/pmd.cache -f xml -auxclasspath lib`

I am wondering if there is a way to speed up the pmd gradle process.  Or the time it takes from the pmd gradle is normal. I want a clean analysis, so I don't want to enable the incremental analysis. 

**Description:**

**Comment by jsotuyod** [comment]
@caesarkim I've a bunch of questions here…

1. which version of Gradle are you using? Can you provide the relevant PMD configuration?
2. what PMD version are you using for stand alone?
3. you say *"I want a clean analysis, so I don't want to enable the incremental analysis."* but you **are** actually using incremental analysis… are we sure the two runs are clean runs and therefore comparable?
4. does the Gradle build include any dependencies beyond the `lib` folder?

**Comment by caesarkim** [comment]
Here is my PMD section in the gradle.

```
subprojects {
	pmd {
		consoleOutput = true
		toolVersion = "6.27.0"
		reportsDir = file("pmd/reports")
		ruleSets = ["${mainDir}/CustomJavaRules.xml"]
		ignoreFailures = true
	}
	tasks.withType(Pmd) {
		aQute.bnd.build.Project bndProject = bndWorkspace.getProject(project.name)
		if (bndProject != null) {
			String[] projectDependencies = bndProject.getDependson()*.getName()
			for (String projectBundle : projectDependencies) {
				project.getDependencies().add('compile', project(':'+projectBundle))
			}
		}
	}
}

```

The version for the standalone PMD - 6.26.0

The gradle does not depend on lib folder. I just had to put all the jars(generated from all the bundles) for standalone PMD to resolve types for the standalone PMD.  I had some issues in the gradle for PMD type resolution, because it couldn't find some jars generated from dependencies, because the jars were built from the bnd plugin (https://plugins.gradle.org/plugin/biz.aQute.bnd.workspace). 

I executed this in the build machine(Jenkins), but I got the same result. With an incremental analysis enabled in my local environment (with the configuration above), it is better. It takes 2 minutes, but it still takes 10 minutes in the build machine. 

Sorry for the confusion about the pmd command line, I compared two different runs(one without cache and the other with cache). Please ignore '-cache C:/Projects/pmd.cache'. The run with cache enabled(from above which I provided) is very fast (about 30 seconds), but we just want a clean analysis for every build. 2 minutes is acceptable, but I am not sure about 10 minutes. 

The only overhead cost that I can think of is the section of adding project dependencies by using bnd. But still I can't imagine that it takes extra 8 minutes for adding dependencies. I don't see 8 minutes  in any other steps (compilation and building jars) prior to the PMD code analysis. There must be something wrong with my configuration. Or any extra step that I need to skip?
